### PR TITLE
core: [refactor] pull Deposed out of Tainted list

### PR DESCRIPTION
**Pull request by phinze**
This is the promised follow up to #1069.

This has us tracking Deposed as a fully separate concept from Tainted in the state, with its own graph transformer and everything.

We also move from single `Eval{Read,Write}State` nodes with lots of parameters to lots of targeted `EvalNodes` that are individually simpler.

### CODE REVIEW

File: terraform/transform_deposed.go [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+				// Then clear the deposed state.
+				&EvalWriteState{
+					Name:         n.ResourceName,
+					ResourceType: n.ResourceType,
+					State:        &emptyState,
+					Deposed:      true,
+				},
+				&EvalReturnError{
+					Error: &err,
+				},
```
Comment (phinze) [inline terraform/transform_deposed.go]: This last four EvalNodes are sort of a mess, and I don't even know if that "always write tainted" works properly. But the tests pass!

File: terraform/transform_tainted.go [hunk 2]
```diff
@@ -43,17 +37,6 @@ func (t *TaintedTransformer) Transform(g *Graph) error {
...
-		if t.Deposed {
-			if t.DeposedInclude {
-				// Only include the deposed resource
-				tainted = rs.Tainted[len(rs.Tainted)-1:]
-			} else {
-				// Exclude the deposed resource
-				tainted = rs.Tainted[:len(rs.Tainted)-1]
-			}
-		}
-
```
Comment (phinze) [inline terraform/transform_tainted.go]: But _this_ `feelsgoodman.gif` :grinning:

File: terraform/eval_error.go [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+// This is useful for scenarios where an error has been captured by
+// another EvalNode (like EvalApply) for special EvalTree-based error
+// handling, and that handling has completed, so the error should be
+// returned normally.
+type EvalReturnError struct {
+	Error *error
+}
+
+func (n *EvalReturnError) Eval(ctx EvalContext) (interface{}, error) {
+	return nil, *n.Error
```
Comment (mitchellh) [inline terraform/eval_error.go]: I would change this to also handle the case where `n.Error` is a nil pointer. At the moment you'll get a panic dereferencing a nil pointer.

File: terraform/state.go [hunk 4]
```diff
@@ -644,6 +653,14 @@ type ResourceState struct {
 	// However, in pathological cases, it is possible for the number
 	// of instances to accumulate.
 	Tainted []*InstanceState `json:"tainted,omitempty"`
+
+	// Deposed is used in the mechanics of CreateBeforeDestroy: the existing
+	// Primary is Deposed to get it out of the way for the replacement Primary to
+	// be created by Apply. If the replacement Primary creates successfully, the
+	// Deposed instance is cleaned up. If there were problems creating the
+	// replacement, we mark the replacement as Tainted and Undepose the former
+	// Primary.
```
Comment (mitchellh) [inline terraform/state.go]: I would just add here for future docs sake that for all intents and purposes, a deposed instance can be treated like a tainted instance in that it must be destroyed.

File: terraform/transform_deposed.go [hunk 5]
```diff
@@ -0,0 +1,157 @@
+package terraform
+
+import "fmt"
+
+// DeposedTransformer is a GraphTransformer that adds tainted resources
+// to the graph.
```
Comment (mitchellh) [inline terraform/transform_deposed.go]: CAUGHT YOU. Fix comment.

File: terraform/transform_deposed.go [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+	// to find deposed resources.
+	View string
+}
+
+func (t *DeposedTransformer) Transform(g *Graph) error {
+	state := t.State.ModuleByPath(g.Path)
+	if state == nil {
+		// If there is no state for our module there can't be any tainted
+		// resources, since they live in the state.
+		return nil
```
Comment (mitchellh) [inline terraform/transform_deposed.go]: Comment issue again

File: terraform/transform_deposed.go [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+	}
+
+	// If we have a view, apply it now
+	if t.View != "" {
+		state = state.View(t.View)
+	}
+
+	// Go through all the resources in our state to look for tainted resources
```
Comment (mitchellh) [inline terraform/transform_deposed.go]: Comment

File: terraform/transform_deposed.go (hunk 1 above)
Comment (mitchellh) [inline terraform/transform_deposed.go]: It sounds like this will work properly now... there might still be a race where we're adding a tainted and a dynamic tainted node is being made, but as you said if its nil nothing will happen.

One case to think about: is it possible to race, fail a destroy, mark it as tainted, then have the tainted node pick it up and try a _second_ destroy. I think we should avoid this.

The easiest thing might just be to just write it back to the deposed.

Comment (phinze) [inline terraform/transform_deposed.go]: Yeah I see what you're saying. I'll take another look at this now that I'm a little more acclimated and after I finish this sanity refactor.

File: terraform/transform_deposed.go (hunk 5 above)
Comment (phinze) [inline terraform/transform_deposed.go]: DANG IT.

**Comment by mitchellh** [comment]
This looks good but there is one big edge case that we should be able to write a test for (and should either way) because we _must_ handle it properly. Assume create before destroy is on:

1. Create resource X2, depose X
2. Ctrl-C.
3. Make change to force a new X2
4. Create X3, depose X2 <== overwrites deposed X

This is because right now we put the deposed transformer onto the `DestroyPrimary` node, and this will always run _after_ the create on CBD. 

What we need to do logically (not sure at this point in the comment how it maps to TF core) is: destroy the deposed first if it exists, then create, then destroy the new deposed if it exists.

Thinking about how to implement... I think the most correct way is to actually introduce a 3rd (ugh) destroy node mode: `DestroyDeposed` (or more accurately, destroy _prior_ deposed), and make sure that the destroy/create BOTH depend on this. i.e. tainted destroys and primary destroys should depend on a deposed destroy.

Any other ideas?

**Comment by phinze** [comment]
Ok the edge case is now handled with 6c93fbb85dfec574d9b5089d7a45b5fc5fb7313d.

Now looking into appropriately wiring in the dependencies like we talked about yesterday, which should be the last outstanding thing before we can merge this sucker.

**Comment by phinze** [comment]
Oh and `eval_state` is begging for another refactor pass...

This blob is like _everywhere_:

```go
	state, lock := ctx.State()

	// Get a read lock so we can access this instance
	lock.RLock()
	defer lock.RUnlock()

	// Look for the module state. If we don't have one, then it doesn't matter.
	mod := state.ModuleByPath(ctx.Path())
	if mod == nil {
		return nil, nil
	}

	// Look for the resource state. If we don't have one, then it is okay.
	rs := mod.Resources[n.Name]
	if rs == nil {
		return nil, nil
	}
```

I'll see if I can either (a) clean it up without too much more time spent on this dang PR, or (b) grit my teeth and mark it for later cleanings up.

**Comment by phinze** [comment]
Alright @mitchellh I THINK IT'S READY. :astonished:

File: terraform/eval_if.go [hunk 8]
```diff
@@ -3,7 +3,8 @@ package terraform
 // EvalIf is an EvalNode that is a conditional.
 type EvalIf struct {
 	If   func(EvalContext) (bool, error)
-	Node EvalNode
+	Then EvalNode
+	Else EvalNode
```
Comment (mitchellh) [inline terraform/eval_if.go]: lol

File: terraform/eval_state.go [hunk 9]
```diff
@@ -5,16 +5,75 @@ import (
...
+		return rs.Primary, nil
+	})
+}
+
+// EvalReadStateTainted is an EvalNode implementation that reads a
+// tainted InstanceState for a specific resource out of the state
+type EvalReadStateTainted struct {
+	Name   string
+	Output **InstanceState
+	// Index indicates which instance in the Tainted list to target, or -1 for the last item.
```
Comment (mitchellh) [inline terraform/eval_state.go]: 80 chars

File: terraform/eval_state.go [hunk 10]
```diff
@@ -5,16 +5,75 @@ import (
...
+}
+
+// Does the bulk of the work for the various flavors of ReadState eval nodes.
+// Each node just provides a reader function to get from the ResourceState to the
+// InstanceState, and this takes care of all the plumbing.
+func readInstanceFromState(
+	ctx EvalContext,
+	resourceName string,
+	output **InstanceState,
+	reader func(*ResourceState) (*InstanceState, error),
```
Comment (mitchellh) [inline terraform/eval_state.go]: Go thing, if its a function type, I usually suffix the var with `Fn`, so: `readerFn` in this case. Its up to you but I find that clearer later in the code to realize I'm calling a function from a var vs a function named `reader.

File: terraform/eval_state.go [hunk 11]
```diff
@@ -221,15 +355,14 @@ func (n *EvalUndeposeState) Eval(ctx EvalContext) (interface{}, error) {
...
 		return nil, nil
 	}
 
-	// Undepose to the tainted
-	idx := len(rs.Tainted) - 1
-	rs.Primary = rs.Tainted[idx]
-	rs.Tainted[idx] = nil
+	// Undepose
+	rs.Primary = rs.Deposed[len(rs.Deposed)-1]
+	rs.Deposed = rs.Deposed[:len(rs.Deposed)-1]
```
Comment (mitchellh) [inline terraform/eval_state.go]: Weird go thing here. We need to do this for GC reasons: `rs.Deposed[idx] = nil` (I did it for tainted before see above). The reason is because the reference is still in the slice, even though its no longer reachable, so the GC will never GC that value until `rs.Deposed` is also completely unreachable.

If you set the value to nil, then it removes the reference.

File: terraform/eval_state_test.go [hunk 12]
```diff
@@ -66,3 +66,164 @@ func TestEvalUpdateStateHook(t *testing.T) {
...
+		Name:         "restype.resname",
+		ResourceType: "restype",
+		State:        &is,
+	}
+	_, err := node.Eval(ctx)
+	if err != nil {
+		t.Fatalf("Got err: %#v", err)
+	}
+
+	rs := state.ModuleByPath(ctx.Path()).Resources["restype.resname"]
```
Comment (mitchellh) [inline terraform/eval_state_test.go]: I find it easier to compare `state.String()`, it is less fragile that way if we change the internal structure of State ever.

**Comment by mitchellh** [comment]
LGTM! WOW.

The read/write state functions are SO MUCH saner. Great job.

File: terraform/eval_state.go (hunk 9 above)
Comment (phinze) [inline terraform/eval_state.go]: Just for this comment... or for the whole file?

File: terraform/eval_state.go (hunk 11 above)
Comment (phinze) [inline terraform/eval_state.go]: Ah yeah, that makes sense. I think this was an artifact of Deposed first being a singleton and then getting flipped to a list. Fixing...

File: terraform/eval_state.go (hunk 10 above)
Comment (phinze) [inline terraform/eval_state.go]: I used `gorename` for the first time to make this change. I think I'm in loooove. :heart_eyes:

**Comment by armon** [comment]
This thread is bringing back nightmares. This was a good idea.
